src/csharp/Intel/Generator/Extended/ExtendedSyntaxGenerator.cs

```diff
@@ -0,0 +1,179 @@
+						}
+						signature.AddArgKind(argKind);
+					}
+				}
+
+				if (toAdd) {
+					var key = new GroupKey(name, signature);
+					if (!groups.TryGetValue(key, out var group)) {
+						group = new OpCodeInfoGroup(name, signature);
+						groups.Add(key, group);
+					}
+					group.Items.Add(code);
+				}
+			}
+
+			var orderedGroups = groups.OrderBy(x => x.Key).Select(x => x.Value).ToArray();
+			Generate(groups, orderedGroups);
+		}
+
+		[DebuggerDisplay("{Name} {Kind}")]
+		protected readonly struct GroupKey : IEquatable<GroupKey>, IComparable<GroupKey> {
+			public GroupKey(string name, Signature signature) {
+				Name = name;
+				Signature = signature;
+			}
+
+			public readonly string Name;
+
+			public readonly Signature Signature;
+
+			public bool Equals(GroupKey other) => Name == other.Name && Signature == other.Signature;
+
+			public override bool Equals(object? obj) => obj is GroupKey other && Equals(other);
+
+			public override int GetHashCode() => HashCode.Combine(Name, Signature);
+
+			public static bool operator ==(GroupKey left, GroupKey right) => left.Equals(right);
+
+			public static bool operator !=(GroupKey left, GroupKey right) => !left.Equals(right);
+
+			public int CompareTo(GroupKey other)
+			{
+				var nameComparison = string.Compare(Name, other.Name, StringComparison.Ordinal);
+				if (nameComparison != 0) return nameComparison;
+				return Signature.CompareTo(other.Signature);
+			}
+		}
+
+		protected struct Signature : IEquatable<Signature>, IComparable<Signature> {
+
+			public int ArgCount;
+
+			ulong _argKinds;
+
+			public ArgKind GetArgKind(int argIndex) => (ArgKind)((_argKinds >> (8 * argIndex)) & 0xFF);
+
+			public void AddArgKind(ArgKind kind) {
+				var shift = (8 * ArgCount);
+				_argKinds = (_argKinds & ~((ulong)0xFF << shift)) | ((ulong)kind << shift);
+				ArgCount++;
+			}
+
+			public override string ToString() {
+				var builder = new StringBuilder();
+				builder.Append('(');
+				for(int i = 0; i < ArgCount; i++) {
+					if (i > 0) builder.Append(", ");
+					builder.Append(GetArgKind(i));
+				}
+				builder.Append(')');
+				return builder.ToString();
+			}
+
+			public bool Equals(Signature other) => ArgCount == other.ArgCount && _argKinds == other._argKinds;
+
+			public override bool Equals(object? obj) => obj is Signature other && Equals(other);
+
+			public override int GetHashCode() => HashCode.Combine(ArgCount, _argKinds);
+
+			public static bool operator ==(Signature left, Signature right) => left.Equals(right);
+
+			public static bool operator !=(Signature left, Signature right) => !left.Equals(right);
+
+			public int CompareTo(Signature other)
+			{
+				var argCountComparison = ArgCount.CompareTo(other.ArgCount);
+				if (argCountComparison != 0) return argCountComparison;
+				return _argKinds.CompareTo(other._argKinds);
+			}
+		}
+		
+		protected enum ArgKind : byte {
+			Unknown,
+			Register,
+			RegisterMemory,
+			Immediate,
+			Immediate8,
```

Comment:
Since imm16/32/64 are all using `Immediate`, is it worth using a `Immediate8` here?

Reply:
I have seen a few instructions where only Immediate8 was necessary and in that case I could enforce the signature with a `sbyte` or something like that... but that just early speculation (that's why I put it in this list for now)... maybe surfacing `long` and checking at runtime the range would be better.

Reply:
I solved this issue differently, so I have only Immediate now.